cmd/gwi-api/main.go

```diff
@@ -0,0 +1,346 @@
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/tchap/gwi-challenge/cmd/gwi-api/api"
+	"github.com/tchap/gwi-challenge/cmd/gwi-api/api/stores/memorystore"
+	"github.com/tchap/gwi-challenge/cmd/gwi-api/api/stores/postgrestore"
+	"github.com/tchap/gwi-challenge/cmd/gwi-api/api/stores/postgrestore/migrations"
+
+	migrate "github.com/golang-migrate/migrate/v4"
+	_ "github.com/golang-migrate/migrate/v4/database/postgres"
+	"github.com/jmoiron/sqlx"
+	"github.com/labstack/echo"
+	"github.com/labstack/echo/middleware"
+	_ "github.com/lib/pq"
+	"github.com/pkg/errors"
+	"github.com/tchap/zapext/types"
+	"go.uber.org/zap"
+	"gopkg.in/urfave/cli.v1"
+)
+
+func main() {
+	app := cli.NewApp()
+	app.Name = "gwi-api"
+	app.Usage = "gwi-api service executable"
+	app.Version = "1.0.0"
+	app.Action = run
+
+	app.Commands = []cli.Command{
+		{
+			Name:  "db",
+			Usage: "manage the database",
+			Subcommands: []cli.Command{
+				// TODO: Would be cool to add more subcommands,
+				//       i.e. to migrate to particular migration step.
+				{
+					Name:   "migrate",
+					Usage:  "migrate the database",
+					Action: runDBMigrate,
+				},
+			},
+		},
+		{
+			Name:    "healthcheck",
+			Aliases: []string{"hc"},
+			Usage:   "run service healthcheck",
+			Action:  runHealthcheck,
+		},
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		log.Fatalf("Error: %+v\n", err)
+	}
+}
+
+func run(c *cli.Context) error {
+	// Load configuration.
+	config, err := LoadConfig()
+	if err != nil {
+		return err
+	}
+
+	// Init logging.
+	logger, err := initLogger(config.DebugEnabled)
+	if err != nil {
+		return err
+	}
+	defer logger.Sync()
+
+	// TODO: Would be cool to add some Prometheus metrics.
+	//       Leaving this as an exercise ;-)
+
+	// XXX: Postgres store (including tests)
+	// XXX: Documentation
+	// XXX: Docker Compose
+
+	// Init the store.
+	store, closeStore, err := initStore(logger, config)
+	if err != nil {
+		return err
+	}
+	defer closeStore()
+
+	// Init the API.
+	a := api.New(logger, store, []byte(config.JWTSecret))
+
+	// Run the healthcheck server in the background.
+	if err := spawnHealthcheckServer(logger, config.HealthcheckPort, a.GetHealthcheck); err != nil {
+		return err
+	}
+
+	// Run the API server.
+	return runAPIServer(logger, config, a)
+}
+
+func initLogger(debugEnabled bool) (*zap.Logger, error) {
+	if debugEnabled {
+		return zap.NewDevelopment()
+	}
+	return zap.NewProduction()
+}
+
+func initStore(logger *zap.Logger, config *Config) (store api.Store, cleanup func(), err error) {
+	if config.DBDisabled {
+		return memorystore.New(), func() {}, nil
+	}
+
+	db, err := sqlx.Open("postgres", config.DBURL)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to open database")
+	}
+
+	return postgrestore.New(logger, db), func() { db.Close() }, nil
+}
+
+func spawnHealthcheckServer(logger *zap.Logger, port int, handler echo.HandlerFunc) error {
+	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", port))
```

Comment:
Shouldn't `127.0.0.1` be replaced by `HTTP_HOST`?

Reply:
It can be, but in our case it was meant as a Docker healthcheck that is only going to ever be checked from within the container, so localhost is cool. But if we wanted to use the endpoint from outside, let's say to call it from a custom monitoring tool, it would have to be replaced.